lib/matplotlib/sphinxext/plot_directive.py

```diff
@@ -312,9 +321,37 @@ def setup(app):
     app.connect('build-finished', _copy_css_file)
     metadata = {'parallel_read_safe': True, 'parallel_write_safe': True,
                 'version': matplotlib.__version__}
+    app.connect('builder-inited', init_filename_registry)
+    app.add_env_collector(_FilenameCollector)
     return metadata
 
 
+# -----------------------------------------------------------------------------
+# Handle Duplicate Filenames
+# -----------------------------------------------------------------------------
+
+def init_filename_registry(app):
+    env = app.builder.env
+    if not hasattr(env, 'mpl_custom_base_names'):
+        env.mpl_custom_base_names = defaultdict(set)
+
+
+class _FilenameCollector(EnvironmentCollector):
+    def process_doc(self, app, doctree):
+        pass
+
+    def clear_doc(self, app, env, docname):
+        if docname in env.mpl_custom_base_names:
+            del env.mpl_custom_base_names[docname]
+
+    def merge_other(self, app, env, docnames, other):
+        for docname in docnames:
```

Comment:
Do we need to loop over the docnames? AFAICS, we only want to merge `other.mpl_custom_base_names` into `env.mpl_custom_base_names`. I.e. we can directly loop over that.

Reply:
OK, I simplified this function. I'm not completely sure about it, but the functionality still seems to work based on manual testing.